internal/ingest/ingest.go

```diff
@@ -623,20 +641,29 @@ func (ing *Ingester) ingestWorker() {
 		select {
 		case <-ing.closeWorkers:
 			return
-		case msg := <-ing.toWorkers:
-			ing.ingestWorkerLogic(msg)
+		case provider := <-ing.toWorkers:
+			ing.ingestWorkerLogic(peer.ID(provider))
 		}
 	}
 }
 
-func (ing *Ingester) ingestWorkerLogic(msg toWorkerMsg) {
+func (ing *Ingester) ingestWorkerLogic(provider peer.ID) {
+
 	// It's assumed that the runIngestStep puts a mutex in this map.
-	ing.providersBeingProcessed[msg.provider].Lock()
-	defer ing.providersBeingProcessed[msg.provider].Unlock()
+	ing.providersBeingProcessed[provider].Lock()
+	defer ing.providersBeingProcessed[provider].Unlock()
+
+	// Pull out the assignment for this provider. Note that runIngestStep populates this atomic.Value.
+	assignmentInterface := ing.providerAdChainStaging[provider].Swap(workerAssignment{none: true})
+	if assignmentInterface == nil || assignmentInterface.(workerAssignment).none {
+		// Nothing to do. Someone took our assignment.
```

Comment:
Please also explain when this could happen.  I am guessing that it can only happen with concurrent sync requests for the same provider and both requests get sent to `ing.toWorkers` before one of them is marked assigned.

Reply:
I don't think this would happen in practice. It's here for completeness.

because of `atomic.Value` even concurrent ingest requests should be fine since we'll know either a worker picked up an assignment and the pending assignment is empty (in which case we properly schedule another one), _or_ we have a pending assignment that we swapped for a more recent assignment and a pending worker hasn't picked it up.

It's not possible to get into a state where two workers are queued on the same assignment, and one takes it and leaves the other with nothing to do.

Unless of course someone manually calls `ingestWorkerLogic` like in a test or something.